File: app/src/main/java/com/hepta/theptavpn/MainActivity.java
package com.hepta.theptavpn;

import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.ServiceConnection;
import android.content.res.ColorStateList;
import android.net.VpnService;
import android.os.Bundle;
import android.os.IBinder;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.content.ContextCompat;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;

import com.hepta.theptavpn.databinding.ActivityMainBinding;

public class MainActivity extends AppCompatActivity {

    private ActivityMainBinding binding;
    MainRecyclerAdapter adapter ;
    private static final int VPN_REQUEST_CODE = 0x0F;
    private LocalVPNService vpnServiceBinder = null;

    public MainViewModel mainViewModel;




    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        binding = ActivityMainBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());
        setTitle(getString(R.string.app_name));
        setSupportActionBar(binding.toolbar);
        binding.fab.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if (mainViewModel.isRunning().getValue()) {
//                    binding.fabProgressCircle.show();
                    stopVpnService();
                    mainViewModel.isRunning().setValue(false);
                }else {
                    mainViewModel.isRunning().setValue(true);

                    startVpnService();

                }
            }
        });

        mainViewModel = new ViewModelProvider(this).get(MainViewModel.class);

        binding.recyclerView.setHasFixedSize(true);
        binding.recyclerView.setLayoutManager(new LinearLayoutManager(this));
        adapter = new MainRecyclerAdapter(this);
        binding.recyclerView.setAdapter(adapter);

        setupViewModel();
        Intent vpnIntent = VpnService.prepare(MainActivity.this);
        if (vpnIntent != null)
            startActivityForResult(vpnIntent, VPN_REQUEST_CODE);
        else
            onActivityResult(VPN_REQUEST_CODE, RESULT_OK, null);

    }

    private void setupViewModel() {
        mainViewModel.getUpdateListAction().observe(this, new Observer<Integer>() {
            @Override
            public void onChanged(Integer index) {
                if (index >= 0) {
                    adapter.notifyItemChanged(index);
                } else {
                    adapter.notifyDataSetChanged();
                }
            }
        });

        mainViewModel.isRunning().observe(this, new Observer<Boolean>() {
            @Override
            public void onChanged(Boolean isRunning) {
                adapter.setRunning(isRunning);
                if (isRunning) {
                    binding.fab.setBackgroundTintList( ColorStateList.valueOf(ContextCompat.getColor(MainActivity.this, R.color.colorSelected)));
                }else {
                    binding.fab.setBackgroundTintList( ColorStateList.valueOf(ContextCompat.getColor(MainActivity.this, R.color.colorUnselected)));
                }

            }
        });
        mainViewModel.isRunning().setValue(false);

    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_main, menu);
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    protected void onResume() {
        super.onResume();
        mainViewModel.reloadServerList();

        Log.e("Rzx","onResume");
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        switch (item.getItemId()){
            case R.id.filter_config:
                Log.e("Rzx","filter_config");
                return true;
            case R.id.add_config:
                Log.e("Rzx","add_config");
                startActivity(new Intent().setClass (MainActivity.this, ConfigActivity.class));
                return true;
        }

        return super.onOptionsItemSelected(item);
    }



    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data)
    {

        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == VPN_REQUEST_CODE && resultCode == RESULT_OK)
        {
            ServiceConnection conn = new ServiceConnection() {
                @Override
                public void onServiceDisconnected(ComponentName name) {
                    vpnServiceBinder = null;
                }
                @Override
                public void onServiceConnected(ComponentName name, IBinder service) {
                    //返回一个MsgService对象
                    vpnServiceBinder = ((LocalVPNService.ProxyBinder)service).getMovieService(MainActivity.this);
                }
            };
//            String guid = mainViewModel.getMainStorage().decodeString(MmkvManager.KEY_SELECTED_SERVER);
            Intent intent = new Intent(getApplicationContext(), LocalVPNService.class);
            bindService(intent, conn, Context.BIND_AUTO_CREATE);
        }
    }
    public void startVpnService(){

        if (vpnServiceBinder ==null){

        }else {
            String guid = mainViewModel.getMainStorage().decodeString(MmkvManager.KEY_SELECTED_SERVER);
            if(guid != null){
                boolean run = vpnServiceBinder.startVpnService(guid);
                Log.e("rzx","run:"+run);
            }else {
                Toast.makeText(this,"select a server",Toast.LENGTH_LONG).show();
            }

        }

//        engine.Key key = new Key();
//        key.setMark(0);
//        key.setMTU(0);
//        key.setDevice("fd://" + tunDevice.getFd()); // <--- here
//        key.setInterface("");
//        key.setLogLevel("debug");
//        key.setProxy("socks5://127.0.0.1:1080"); // <--- and here
//        key.setRestAPI("");
//        key.setTCPSendBufferSize("");
//        key.setTCPReceiveBufferSize("");
//        key.setTCPModerateReceiveBuffer(false);
//
//        engine.Engine.insert(key);
//        engine.Engine.start();



    }
    public void stopVpnService() {
        vpnServiceBinder.stopVpnService();
    }

}

File: app/src/main/java/com/hepta/theptavpn/MainActivity.kt
package com.hepta.theptavpn

import android.Manifest
import android.app.*

import android.content.ComponentName
import android.content.Intent
import android.content.ServiceConnection
import android.content.pm.PackageManager
import android.content.res.ColorStateList
import android.graphics.Color
import android.net.VpnService
import android.os.Build
import android.os.Bundle
import android.os.IBinder
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.widget.Toast
import androidx.activity.result.ActivityResult
import androidx.activity.result.ActivityResultLauncher
import androidx.activity.result.contract.ActivityResultContracts
import androidx.activity.viewModels
import androidx.appcompat.app.ActionBarDrawerToggle
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import androidx.core.app.NotificationCompat
import androidx.core.content.ContextCompat
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import com.google.android.material.navigation.NavigationView
import com.hepta.theptavpn.LocalVPNService.ProxyBinder
import com.hepta.theptavpn.databinding.ActivityMainBinding


class MainActivity : AppCompatActivity(), NavigationView.OnNavigationItemSelectedListener {
    private var binding: ActivityMainBinding? = null
    var adapter: MainRecyclerAdapter? = null
    private var localVPNService: LocalVPNService? = null
    private var manager: NotificationManager? = null
    private var note: Notification? = null
    val mainViewModel: MainViewModel by viewModels()



    private val notificationRequestPermissionLauncher: ActivityResultLauncher<String> = registerForActivityResult(ActivityResultContracts.RequestPermission()){}


    private val startForResult = registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { result: ActivityResult ->
        if (result.resultCode == Activity.RESULT_OK) {
            startVpnService()
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        Log.e("Rzx","onCreate")
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding!!.root)
        title = getString(R.string.app_name)
        setSupportActionBar(binding!!.toolbar)
        binding!!.fab.setOnClickListener {
            if (mainViewModel!!.isRunning.value!!) {
                stopVpnService()
            } else {
                val vpnIntent = VpnService.prepare(this@MainActivity)
                if (vpnIntent != null){
                    startForResult.launch(vpnIntent)
                }else{
                    startVpnService()
                }
            }
        }
        binding!!.recyclerView.setHasFixedSize(true)
        binding!!.recyclerView.layoutManager = LinearLayoutManager(this)
        adapter = MainRecyclerAdapter(this)
        binding!!.recyclerView.adapter = adapter
        val toggle = ActionBarDrawerToggle(
            this,
            binding!!.drawerLayout,
            binding!!.toolbar,
            R.string.navigation_drawer_open,
            R.string.navigation_drawer_close
        )
        binding!!.drawerLayout.addDrawerListener(toggle)
        toggle.syncState()
        binding!!.navView.setNavigationItemSelectedListener(this)
        binding!!.version.text = BuildConfig.VERSION_NAME
        setupViewModel()
        createNotification()
//        requestPermission()
    }


    private fun requestPermission() {

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            if (lacksPermission(Manifest.permission.POST_NOTIFICATIONS)){
                if (ActivityCompat.shouldShowRequestPermissionRationale(this, Manifest.permission.POST_NOTIFICATIONS)) {
                    deniedPermissionDialog("Notification Permission Required", " App need notification permission to enable this feature.")
                } else {
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
                        notificationRequestPermissionLauncher.launch(Manifest.permission.POST_NOTIFICATIONS)
                    }
                }
            }
        }
    }

    /**
     * 判断是否缺少权限
     */
    private fun lacksPermission( permission: String): Boolean {
        return ContextCompat.checkSelfPermission(applicationContext, permission) ==
                PackageManager.PERMISSION_DENIED
    }


    private fun deniedPermissionDialog(title: String, message: String) {
        val builder = AlertDialog.Builder(this)
        builder.setTitle(title)
            .setMessage(message)
            .setPositiveButton("OK") { dialogInterface, i -> dialogInterface.dismiss() }
        builder.create().show()
    }


    private fun createNotification() {
        manager = getSystemService(NOTIFICATION_SERVICE) as NotificationManager

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val channel = NotificationChannel(
                getString(R.string.notify_channel_id), getString(R.string.notify_channel_name),
                NotificationManager.IMPORTANCE_DEFAULT
            )
            manager!!.createNotificationChannel(channel)
        }
        val intent = Intent(this, MainActivity::class.java)
        val pending = PendingIntent.getActivity(this, 0, intent, PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE)
        note = NotificationCompat.Builder(this, "ShadyPi")
            .setContentTitle("Thepta Application Capture Packet")
            .setContentText("running").setContentIntent(pending)
            .setSmallIcon(R.drawable.ic_launcher_foreground).setOngoing(true)
            .setColor(Color.parseColor("#ff0000")) //设置小图标颜色
            .setAutoCancel(false)
            .build()
    }


    private fun cancelNotification() {
        manager?.cancel(1);


    }
    private fun showNotification() {
        manager?.notify(1,note);

    }


    private fun setupViewModel() {
        mainViewModel!!.updateListAction.observe(this) { index ->
            if (index >= 0) {
                adapter!!.notifyItemChanged(index)
            } else {
                adapter!!.notifyDataSetChanged()
            }
        }
        mainViewModel!!.isRunning.observe(this) { isRunning ->
            adapter!!.isRunning = isRunning
            if (isRunning) {
                binding!!.fab.backgroundTintList = ColorStateList.valueOf(ContextCompat.getColor(this@MainActivity, R.color.colorSelected))
                showNotification()
            } else {
                binding!!.fab.backgroundTintList = ColorStateList.valueOf(ContextCompat.getColor(this@MainActivity, R.color.colorUnselected))
                cancelNotification()
            }
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_main, menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onResume() {
        super.onResume()
        mainViewModel!!.reloadServerList()
        if(localVPNService==null){
            val conn: ServiceConnection = object : ServiceConnection {
                override fun onServiceDisconnected(name: ComponentName)
                {
                    localVPNService = null
                }
                override fun onServiceConnected(name: ComponentName, service: IBinder)
                {
                    //返回一个MsgService对象
                    localVPNService = (service as ProxyBinder).getLocalVPNService(this@MainActivity)
                    mainViewModel.isRunning.value= localVPNService!!.Running
                }
            }
            val intent = Intent(applicationContext, LocalVPNService::class.java)
            bindService(intent, conn, BIND_AUTO_CREATE)
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.add_config -> {
                startActivity(Intent().setClass(this@MainActivity, ConfigActivity::class.java))
                return true
            }
        }
        return super.onOptionsItemSelected(item)
    }


    fun startVpnService() {
        val guid = mainViewModel!!.mainStorage.decodeString(MmkvManager.KEY_SELECTED_SERVER)
        if (guid != null) {
            localVPNService?.startVpnService(guid)

        } else {
            Toast.makeText(this, "select a server", Toast.LENGTH_LONG).show()
        }

    }

    fun stopVpnService() {
        localVPNService?.stopVpnService()

    }

    override fun onNavigationItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.settings -> {
                startActivity(Intent(this, NSettingsActivity::class.java))
            }
        }
        return false
    }
}